### DISCUSSION TITLE
Aysnc Dispatch for MPMD?

### DISCUSSION THREAD

**Discussion by YigitElma** [jax-ml/jax#34013] [Q&A]
Hi,

I am working on a multi-objective optimization problem that is both compute and memory-intensive. Each objective involves custom functions and custom data. As the problem scales, the full state and computation cannot fit on a single GPU. This is effectively a Multiple Program Multiple Data (MPMD) problem: I need to run Function A on GPU 0 and Function B on GPU 1 simultaneously.

I have explored several ways to distribute this workload last year, but each has drawbacks:
- Python Multiprocessing: Unfortunately, the fork option didn't play well with JAX and the spawn/forkserver options had substantial overhead.
- Python Threading: I couldn't get much success with this.
- MPI: I can accomplish everything I need with MPI, however the code is fairly complex and hard to maintain. Also, the data movement is probably not ideal.

The MPI option is currently working as expected. However, for the user, it creates a significant complexity to use the otherwise very simple code. I am once again looking for the available options, and since the sharding interface got significantly better in the last year, I thought maybe a pure JAX version is possible?

I naively think that JAX's asynchronous dispatch can help to achieve a pseudo-MPMD, such that if I have multiple functions jitted to run on different devices, when I call multiple of these (all of them with different device), beside the sequential CPU dispatch, all the GPUs will be working simultaneously at some point? For example,

```python
@partial(jax.jit, device=jax.devices("gpu")[0])
def fun1():
    # do some heavy computation on GPU 0

@partial(jax.jit, device=jax.devices("gpu")[1])
def fun2():
    # do some heavy computation on GPU 1

@partial(jax.jit, device=jax.devices("gpu")[2])
def fun3():
    # do some heavy computation on GPU 2

x1 = some_data_on_gpu0
x2 = some_data_on_gpu1
x3 = some_data_on_gpu2

out = jnp.concatenate([fun1(x1), fun2(x2), fun3(x3)])
```

As I said, I naively think the final line will use all the GPUs simultaneously after the CPU is done dispatching the workload. Is this possible? Or is there another way to achieve the same goal purely in JAX without MPI?

Related to #6678 and #17356

**Comment by Shekar-77** [comment]
https://docs.jax.dev/en/latest/multi_process.html you can refer this documentation for understanding how parallel programming works in jax.

**Comment by DiagRisker** [comment]
Hello @YigitElma ,

The link provided by @Shekar-77 will help however some infos I found were not there. I started a little topic on the matter a while ago : https://github.com/jax-ml/jax/discussions/32494#discussioncomment-14885451
For automatic sharding setup (as long as you jit your tensor dispatch **within a jitted function**), you can use : 
```python
mesh = Mesh( jax.devices(), 
        axis_names = ('0th axis'), 
        axis_types =(jax.sharding.AxisType.Auto,),
        ) # jax.make_mesh is another option
jax.sharding.set_mesh(mesh)
```
this will dispatch tasks and data.

if you want a precise dispatch (as mentionned in the tuto I linked),  you can jit a function with:
```python
sharder=NamedSharding(mesh, jax.P(('0th axis'))) 
@jax.jit( in_shardings = sharder, out_shardings = sharder )
```
A general guide can be found here too: https://docs.jax.dev/en/latest/notebooks/shard_map.html
but be aware, that the explicit sharding requires you to know the precise tensor order and indexing --- you can't just allocate a precise axis for the parallel dispatch to occur, you need to state every bit of the tensor in the axis bit of jax.sharding.PartitionSpec (or jax.P), which is less handy than cases were you have jax.numpy.sum(..., axis = <mylist of axis to contract (the rest being resolved automatically, where sharding does not))> ) .
This is the limitation I found (feel free to correct me)
I did not successfully test `@partial(jax.jit, device=jax.devices("gpu")) ` on my setup but it should be easier to work with.

An advanced example can be found in recent forums discussions: https://github.com/jax-ml/jax/discussions/34076 ( but it involves threads from python3.14t)

**Comment by YigitElma** [comment]
Thanks @Shekar-77 and @DiagRisker for the replies!
Unfortunately, [JAX Multi-host tutorial](https://docs.jax.dev/en/latest/multi_process.html) is not what I am looking for. As I described above, I want to do Multiple Program Multiple Data, not SPMD. Tutorial mentions that it is for SPMD,
<img width="799" height="374" alt="image" src="https://github.com/user-attachments/assets/6edd30f1-e01a-493e-951c-3304f949fbf4" />


#34076 is closer to what I want, but I am not sure if the `Threading` is the best approach in my case. However, I would like to hear more from people that use `Threading` for these kind of applications, such as single CPU with multiple GPUs, and multiple CPUs with just 1 process.

**Comment by seanprime7** [comment]
Have you considered using [JaxPP](https://github.com/NVIDIA/jaxpp)?
It appears that your use case may benefit from it.

**Reply by YigitElma** [reply]
> Have you considered using [JaxPP](https://github.com/NVIDIA/jaxpp)? It appears that your use case may benefit from it.

Hi @seanprime7, this looks very interesting. I will definitely take a look at it! Thanks!
